packages/eslint-plugin/src/rules/no-redundant-type-constituents.ts

```diff
@@ -323,8 +327,58 @@ export default createRule({
               }
             }
           }
-        }
 
+          typePartFlags.length >= 2
+            ? seenUnionTypes.set(typeNode, typePartFlags)
+            : null;
+        }
+        const checkIfUnionsAreAssignable = (): undefined => {
+          let typeFlagsOfUnions: ts.TypeFlags[] = [];
+          let result = true;
+          let primitiveUnit: number;
+
+          seenUnionTypes.forEach((value, key) => {
+            value.forEach(union => {
+              typeFlagsOfUnions.push(union.typeFlags);
+            });
+            for (const iterator of typeFlagsOfUnions) {
+              if (
+                seenPrimitiveTypes.has(
+                  literalToPrimitiveTypeFlags[
+                    iterator as keyof typeof literalToPrimitiveTypeFlags
+                  ],
+                )
+              ) {
+                result = true;
+                primitiveUnit =
+                  literalToPrimitiveTypeFlags[
+                    iterator as keyof typeof literalToPrimitiveTypeFlags
+                  ];
```

Comment:
[Question] Here, `primitiveUnit` is set to its primitive type flags number. But then in the `context.report` call, it's used as a key for the type flag names. It seems a bit 'extra' to me to have two places where lookups are being done. How about unifying and renaming to a more succinct name like `primitive`?

```suggestion
                primitive = literalToPrimitiveTypeFlags[iterator];
```